Creating datatypes.Stack  of struct references produces C error.

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Manjaro Linux"
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8350U CPU @ 1.70GHz
CC version: cc (GCC) 12.1.0

vexe mtime: 2022-06-21 18:28:08
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 223b96a.8172fec

Git version: git version 2.36.1
Git vroot status: weekly.2022.24-52-g8172fecb-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 eee03626

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg test.v`
```v
import datatypes { Stack }

[heap]
struct Element {
mut:
	name       string
	value      string
}

fn main() {
	mut parent := &Element{
		name: 'parent element'
	}
	mut stack := Stack<&Element>{}
}

```

**What did you expect to see?**

Expected stack with generic type of a struct reference to compile


**What did you see instead?**
```
test.v:11:6: warning: unused variable: `parent`
    9 | 
   10 | fn main() {
   11 |     mut parent := &Element{
      |         ~~~~~~
   12 |         name: 'parent element'
   13 |     }
test.v:14:6: warning: unused variable: `stack`
   12 |         name: 'parent element'
   13 |     }
   14 |     mut stack := Stack<&Element>{}
      |         ~~~~~
   15 | }
==================
/usr/lib64/crt1.o: error: Invalid relocation entry [15] '.rela.debug_info' @ 000004f5
tcc: error: file 'crt1.o' not found
/tmp/v_1000/test.11332462667088156692.tmp.c:89: error: ';' expected (got "*")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1161,7 +1161,7 @@ fn (mut g Gen) option_type_name(t ast.Type) (string, string) {
 	} else {
 		styp = '${c.option_name}_${base}'
 	}
-	if t.is_ptr() {
+	if t.is_ptr() || t.has_flag(.generic) {
 		styp = styp.replace('*', '_ptr')
 	}
 	return styp, base
@@ -1183,7 +1183,7 @@ fn (mut g Gen) result_type_name(t ast.Type) (string, string) {
 	} else {
 		styp = '${c.result_name}_${base}'
 	}
-	if t.is_ptr() {
+	if t.is_ptr() || t.has_flag(.generic) {
 		styp = styp.replace('*', '_ptr')
 	}
 	return styp, base

--- a/vlib/v/tests/option_ret_ptr_generic_test.v
+++ b/vlib/v/tests/option_ret_ptr_generic_test.v
@@ -1,0 +1,30 @@
+struct Stack[T] {
+mut:
+	elements []T
+}
+
+pub fn (stack Stack[T]) is_empty() bool {
+	return stack.elements.len == 0
+}
+
+pub fn (stack Stack[T]) peek() ?T {
+	return if !stack.is_empty() { stack.elements.last() } else { none }
+}
+
+pub fn (stack Stack[T]) peek2() !T {
+	return if !stack.is_empty() { stack.elements.last() } else { error('Stack is empty') }
+}
+
+@[heap]
+struct Element {
+mut:
+	name  string
+	value string
+}
+
+fn test_main() {
+	mut parent := &Element{
+		name: 'parent element'
+	}
+	mut stack := Stack[&Element]{}
+}

